Guide JAX-RS

hollingsworthd opened the pull request:
To build the guide without executing the tests run:

` ./gradlew -x micronautJaxrsJdbcRunTestScript micronautJaxrsJdbcBuild --console=plain
`
`open open build/dist/index.html
`
To run and exectute the tests:

` ./gradlew micronautJaxrsJdbcBuild --console=plain`

sdelamo committed:
jax-rs guide

To build the guide without executing the tests run:

 ./gradlew -x micronautJaxrsJdbcRunTestScript micronautJaxrsJdbcBuild --console=plain

open open build/dist/index.html

To run and exectute the tests:

 ./gradlew micronautJaxrsJdbcBuild --console=plain

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/main/java/example/micronaut/Application.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Override
+    public void onApplicationEvent(StartupEvent event) { // <2>
+        if (LOG.isInfoEnabled()) {
```
no need for enabled check

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/main/java/example/micronaut/Application.java:
```diff
@@ -0,0 +1,53 @@
+    private final PetRepository petRepository;
+
+    public static void main(String[] args) {
+        Micronaut.run(javax.ws.rs.core.Application.class, args);
```
is `javax.ws.rs.core.Application` correct?

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/main/java/example/micronaut/domain/Owner.java:
```diff
@@ -0,0 +1,44 @@
+
+    private int age;
+
+    @Creator
```
no need for @Creator if there's only 1 constructor

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/main/java/example/micronaut/domain/Pet.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    public PetType getType() {
+		return type;
```
please fix formatting

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/main/resources/db/migration/V1__schema.sql:
```diff
@@ -0,0 +1,15 @@
+// DDL for an H2 database
```
use # for SQL comment

hollingsworthd replied:
H2 requires `//`

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/main/resources/logback.xml:
```diff
@@ -0,0 +1,15 @@
+<configuration>
```
delete this file, it's generated

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/test/java/example/micronaut/MicronautguideTest.java:
```diff
@@ -0,0 +1,22 @@
+package example.micronaut;
```
is this needed? the only difference between this and the generated class is `@MicronautTest(application = Application.class)` - if that's needed it should have a callout and explanation

burtbeckwith commented:
When I ran the `mn create-app` command it failed: `Error The requested feature does not exist: jakarta-javax-persistence`. Also `#tag::flyway[]` and `#end::flyway[]` end up in the rendered HTML

hollingsworthd commented:
Thanks @burtbeckwith I think I've addressed the feedback.

Also I added some additional text to the guide body to introduce the classes.

Commit:
Add JAX-RS documentation (#866)

sdelamo committed:
jaxrs 101 guide

sdelamo commented:
@kroegerk can you proofread?

burtbeckwith commented on src/docs/common/snippets/common-requirements-mysql-testcontainers.adoc:
```diff
@@ -0,0 +1,2 @@
+common:requirements.adoc[]
+* https://www.docker.io/gettingstarted/#h_installation[Docker] installed to run MySQL and to run tests using https://www.testcontainers.org[Testcontainers].
```
this would be more reusable if `MySQL` was a parameter

burtbeckwith commented on src/docs/common/snippets/common-datasource-configuration.adoc:
```diff
@@ -0,0 +1,7 @@
+=== Data Source configuration
+
+Define the datasource in `src/main/resources/application.yml`.
```
this is wrong and shouldn't be a common snippet like this. there's already a generated datasource, so you should have them replace it with the one from the guide, not generate it

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/metadata.json:
```diff
@@ -0,0 +1,14 @@
+  "authors": ["Dan Hollingsworth"],  
+  "category": "API",
+  "publicationDate": "2022-03-22",
+  "languages": ["java"],
```
why no Groovy or Kotlin?

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/metadata.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "title": "Micronaut JAX-RS",
```
create a proper title and intro

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/micronaut-jaxrs-jdbc.adoc:
```diff
@@ -0,0 +1,94 @@
+
+common:datasource-configuration.adoc[]
+
+common:flyway.adoc[]
```
shouldn't use this include - the flyway dependency is already there since it was specified in the features. all that's needed is the part about enabling it in application.yml

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/main/resources/db/migration/V1__schema.sql:
```diff
@@ -0,0 +1,7 @@
+DROP TABLE IF EXISTS pet;
+
+CREATE TABLE pet (
+                     id   BIGINT NOT NULL AUTO_INCREMENT UNIQUE PRIMARY KEY,
```
indent with 4 spaces, not _21_

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/main/java/example/micronaut/Pet.java:
```diff
@@ -0,0 +1,63 @@
+package example.micronaut;
+
+import io.micronaut.core.annotation.Generated;
```
unused

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/main/java/example/micronaut/NameDto.java:
```diff
@@ -0,0 +1,24 @@
+import io.micronaut.core.annotation.NonNull;
+
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.NotNull;
```
unused

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/test/java/example/micronaut/PetResourceTest.java:
```diff
@@ -0,0 +1,85 @@
+
+import java.net.URI;
+import java.util.Arrays;
+import java.util.Collections;
```
unused

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/test/java/example/micronaut/PetResourceTest.java:
```diff
@@ -0,0 +1,85 @@
+import java.util.Collections;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
```
no star imports

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/main/java/example/micronaut/PetResource.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    @GET // <3>
+    List<NameDto> all() { // <4>
```
this and `byName` work in Micronaut because the generated bean class is in the same package, but IntelliJ complains that the two methods should be public. I'd make the scope `public` so readers don't think there's a real problem

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/test/java/example/micronaut/PetResourceTest.java:
```diff
@@ -0,0 +1,85 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@MicronautTest(transactional = false) // <1>
```
tell them why you're not using transactions

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/micronaut-jaxrs-jdbc.adoc:
```diff
@@ -0,0 +1,94 @@
+common:testApp.adoc[]
+
+common:runapp-mysql-with-docker.adoc[]
+
```
you have them start the app but don't have them use it. should create some data and view it at the least

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/micronaut-jaxrs-jdbc.adoc:
```diff
@@ -0,0 +1,94 @@
+common:runapp-mysql-with-docker.adoc[]
+
+common:graal-with-plugins.adoc[]
+
```
repeat the usage info (TBD) from the run app section here so they can see that everything works that same with native image

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/micronaut-jaxrs-jdbc.adoc:
```diff
@@ -0,0 +1,94 @@
+common:header.adoc[]
+
+The application exposes some REST endpoints using https://projects.eclipse.org/projects/ee4j.jaxrs[JAX-RS annotations] and stores data in a MySQL database using Micronaut Data JDBC.
```
need to explain what JAX-RS is, why they'd use it, why they'd use it instead of what's available in Micronaut, point out the micronaut-jaxrs docs, etc.

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/micronaut-jaxrs-jdbc.adoc:
```diff
@@ -0,0 +1,94 @@
+
+The above entity uses an enum:
+
+source:PetType[]
```
this should be before the Pet class so Pet compiles when it's added

burtbeckwith commented on src/docs/common/snippets/common-gettingStarted.adoc:
```diff
@@ -1,3 +1,5 @@
 == Getting Started
 
 In this guide, we will create a Micronaut application written in @language@.
+
```
remove these unnecessary blank lines

sdelamo committed:
move enum before entity

sdelamo committed:
remove unused NotNull

sdelamo committed:
remove unused Generated

sdelamo committed:
use 4 spaces

sdelamo committed:
remove unused import

sdelamo committed:
don't use wildcard imports

sdelamo committed:
better intro

sdelamo committed:
change copy of common flyway

from:

Add the following snippet to include the necessary dependencies:

to:

The `flyway` feature adds the following dependency:

sdelamo committed:
remove unncessary blank lines

sdelamo committed:
instructions to run the app

sdelamo committed:
explain guides goal

kroegerk commented on guides/micronaut-jaxrs-jdbc/micronaut-jaxrs-jdbc.adoc:
```diff
@@ -0,0 +1,164 @@
+By default, Micronaut users define their https://docs.micronaut.io/latest/guide/#routing[HTTP Routing] using the Micronaut `@Controller` annotation and other built-in https://docs.micronaut.io/latest/guide/#_routing_annotations[Routing Annotations]. However, https://micronaut-projects.github.io/micronaut-jaxrs/latest/guide/[Micronaut JAX-RS] allows you to define your Micronaut endpoints with JAX- ...
+
+____
+What exactly is JAX-RS? JAX-RS is a POJO based, annotation driven framework for building web services which comply with RESTful principles. Imagine writing all the low level code to parse a HTTP request and the logic just to wire these requests to appropriate Java classes/methods. The beauty of the JAX-RS API is that it insulates the developer from the complexity and allows them to concentrate on ...
```
```suggestion
What exactly is JAX-RS? JAX-RS is a POJO-based, annotation-driven framework for building web services that comply with RESTful principles. Imagine writing all the low level code to parse an HTTP request and the logic just to wire these requests to appropriate Java classes/methods. The beauty of the JAX-RS API is that it insulates developers from that complexity and allows them to concentrate on business logic. That’s precisely where the use of POJOs and annotations come into play! JAX-RS has annotations to bind specific URI patterns and HTTP operations to individual methods of your Java class.
```

kroegerk commented on guides/micronaut-jaxrs-jdbc/micronaut-jaxrs-jdbc.adoc:
```diff
@@ -0,0 +1,164 @@
+
+==== Dependencies
+
+When you add `jax-rs` feature, the generated application includes the following dependencies:
```
```suggestion
When you add a `jax-rs` feature, the generated application includes the following dependencies:
```

burtbeckwith replied:
should be `the`, not `a`

kroegerk commented on guides/micronaut-jaxrs-jdbc/micronaut-jaxrs-jdbc.adoc:
```diff
@@ -0,0 +1,164 @@
+
+<1> A JAX-RS annotation to define the base URI for methods in this class.
+callout:constructor-di[number=2,arg0=PetRepository]
+<3> A JAX-RS annotation to define an endpoint for HTTP Get requests. Mapped to the base URI `/pets` since no `@Path` is provided.
```
```suggestion
<3> A JAX-RS annotation to define an endpoint for HTTP Get requests. Mapped to the base URI `/pets`, since no `@Path` is provided.
```

kroegerk commented on guides/micronaut-jaxrs-jdbc/micronaut-jaxrs-jdbc.adoc:
```diff
@@ -0,0 +1,164 @@
+<1> A JAX-RS annotation to define the base URI for methods in this class.
+callout:constructor-di[number=2,arg0=PetRepository]
+<3> A JAX-RS annotation to define an endpoint for HTTP Get requests. Mapped to the base URI `/pets` since no `@Path` is provided.
+<4> The response is converted to JSON Array automatically.
```
```suggestion
<4> The response is converted to a JSON array automatically.
```

kroegerk commented on guides/micronaut-jaxrs-jdbc/micronaut-jaxrs-jdbc.adoc:
```diff
@@ -0,0 +1,164 @@
+callout:constructor-di[number=2,arg0=PetRepository]
+<3> A JAX-RS annotation to define an endpoint for HTTP Get requests. Mapped to the base URI `/pets` since no `@Path` is provided.
+<4> The response is converted to JSON Array automatically.
+<5> Paths can be templated for path parameters, such as the pet's name here.
```
```suggestion
<5> Paths can be templated for path parameters, such as the pet's name, here.
```

kroegerk commented on guides/micronaut-jaxrs-jdbc/micronaut-jaxrs-jdbc.adoc:
```diff
@@ -0,0 +1,164 @@
+
+callout:micronaut-test-transactional-false[1]
+callout:http-client[2]
+<3> Micronaut returns a 404 response if the resource method returns an empty optional
```
```suggestion
<3> Micronaut returns a 404 response if the resource method returns an empty optional.
```

kroegerk commented on src/docs/common/snippets/common-datasource-configuration.adoc:
```diff
@@ -0,0 +1,7 @@
+
+resource:application.yml[tag=datasource]
+
+NOTE: This way of defining the datasource properties enables us to externalize the configuration, for example for production environment, and also provide a default value for development. If the environment variables are not defined, the Micronaut framework will use the default values.
```
```suggestion
NOTE: This way of defining datasource properties enables us to externalize the configuration, for example for a production environment, and also provide a default value for development. If the environment variables are not defined, the Micronaut framework will use the default values.
```

Commit:
Address code review feedback

sdelamo committed:
Update guides/micronaut-jaxrs-jdbc/micronaut-jaxrs-jdbc.adoc

Co-authored-by: Kristin Kroeger <33693760+kroegerk@users.noreply.github.com>

sdelamo committed:
Update src/docs/common/snippets/common-datasource-configuration.adoc

Co-authored-by: Kristin Kroeger <33693760+kroegerk@users.noreply.github.com>

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/metadata.json:
```diff
@@ -0,0 +1,14 @@
+  "apps": [
+    {
+      "name": "default",
+      "features": ["jax-rs", "data-jdbc", "mysql", "flyway", "jdbc-hikari", "graalvm"]
```
can remove `jdbc-hikari`, it's pulled in by `data-jdbc`

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/main/java/example/micronaut/PetSave.java:
```diff
@@ -0,0 +1,34 @@
+import javax.validation.constraints.NotNull;
+
+@Introspected
+public class PetSave {
```
minor, but the name seems weird to me. maybe call it PetSaveCommand?

burtbeckwith commented on guides/micronaut-jaxrs-jdbc/java/src/test/java/example/micronaut/PetResourceTest.java:
```diff
@@ -0,0 +1,100 @@
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+@MicronautTest(transactional = false) // <1>
```
still don't know _why_ the test isn't transactional

hollingsworthd replied:
I'm planning to look into this more. This code was originally from the examples repo. I appreciate your thorough review, Bert and will be spending some more time with this PR in the coming days.

Commit:
Address code review feedback

sdelamo committed:
add @MicronautTest(transactional = false) back

sdelamo committed:
external common-datasource-configuration

sdelamo committed:
rename to PetSave otherwise it does not compile

sdelamo committed:
add twitter card

sdelamo committed:
add sergio as author

sdelamo committed:
set publication date to May 5th

sdelamo committed:
Merge branch 'master' into jax-rs-docs

sdelamo committed:
Merge branch 'master' into jax-rs-docs
